knimin/lib/data_access.py

```diff
@@ -827,6 +827,104 @@ def pulldown(self, barcodes, blanks):
 
         return metadata, failures
 
+    def explain_pulldown_failures(self, barcodes):
+        """Builds failure reason list for barcodes passed
+
+        Parameters
+        ----------
+        barcodes : list of str
+            Barcodes to explain failure for
+
+        Returns: list of tuples of str
+            list in the form [(barcode, reason), (barcode, reason), ...]
+        """
+        def update_failure(sql, all_bc, reason):
+            hold = [x[0] for x in
+                    self._con.execute_fetchall(sql, [tuple(all_bc)])]
+            remaining_bc = all_bc.difference(hold)
+            return [(bc, reason) for bc in hold], remaining_bc
+
+        failure_info = []
+        current_barcodes = set(barcodes)
+
+        # TEST ORDER HERE MATTERS!
```

Comment:
It's a cascade of tests, so removal order matters, e.g. first it removes all non-AG barcodes so you can assume the remaining ones belong to AG. Then it removes handout ones, so you can assume the rest are assigned to registered kits, etc.